magicicadaprotocol/proxy_tunnel.py

```diff
@@ -66,18 +70,25 @@ def connectionMade(self):
         """Connection was established."""
         self.negotiation_done = False
         self.client_protocol = None
-        self.__buffer = ""
+        self.__buffer = b""
         # send request
-        line = "CONNECT %s:%s HTTP/1.0\r\n"
-        self.transport.write(line % (self.factory.host, self.factory.port))
+        line = "CONNECT %s:%s HTTP/1.0" % (
+            self.factory.host, self.factory.port)
+        self.write_line(line)
         # send headers
         self.sendHeader("Host", self.factory.host)
         # do auth
         user = self.factory.user
         if user is not None:
-            auth_str = base64.b64encode("%s:%s" % (user, self.factory.passwd))
-            self.sendHeader("Proxy-Authorization", "Basic %s" % auth_str)
-        self.transport.write("\r\n")
+            if isinstance(user, str):
+                user = user.encode('utf8')
+            passwd = self.factory.passwd
+            if isinstance(passwd, str):
+                passwd = passwd.encode('utf8')
```

Comment:
It's weird that this (and `user` above) could be `str` some times, `bytes` others...

Reply:
So the key here is that the code needs those to be bytes, because the concatenation is then b64encoded. One option I considered is to force the values to be either str or bytes, and then adjust the callers to do the right thing. That felt a little less backward compatible so in just this one piece of code I considered the check and optional encode to be the less intrusive change.